NAS-106037 / 12.0 / Make MTU of VNet interfaces configurable

nathanjrobertson opened the pull request:
This patch adds five new configuration parameters, with the default behaviour being identical to the pre-existing behaviour (copy the MTU value from the bridge, or 1500 if there is no existing bridge). We add:

* vnet[0-3]_mtu - default to "auto" (ie. existing behaviour). If set to a value, it will take top precedence in creating the device (ie. it will ignore the value the bridge has)
* vnet_default_mtu - replaces the hardcoded "1500" value in the code with a configurable variable (but the variable defaults to "1500").

The use case we have for this where a bridge (say bridge0) has an IP address, and we bring it up on boot with that address (but no bridged NIC attached). if_bridge(4) says that you can't set the MTU on bridge0 with ifconfig - it is configured from the first attached device to the bridge. In our case, the VNet device ends up being the first attached, and this is hardcoded to 1500 (the VPS provider the machine runs on requires it to be 1450). Using one of the two above options (it seemed sensible to provide either a global or per-interface configuration) covers this situation.

nathanjrobertson committed:
Add configuration parameters for default MTU on VNet device creation, as well as specific MTU for each VNet device. Default behaviour is the old behaviour (default of 1500, or whatever the bridge we're connecting to is set to)

nathanjrobertson committed:
Merge branch 'master' of https://github.com/iocage/iocage

nathanjrobertson committed:
Add a space after a comma (travis fussiness)

skarekrow reviewed (dismissed):
LGTM, @sonicaj?

nathanjrobertson committed:
Mock a hard code variable for unit test use only

nathanjrobertson commented:
@skarekrow Tests now pass. Could you please review again? Thanks.

skarekrow reviewed (dismissed):
Still LGTM

nathanjrobertson commented:
@sonicaj Any feedback? I'd very much like this patch to make the next iocage release, as we're using it in production currently.

nathanjrobertson committed:
Merge https://github.com/iocage/iocage

nathanjrobertson commented:
@skarekrow I had to merge a conflict in iocage_json.py, as there was another configuration option added which took the version number I was using. No other changes, just a merge. Could you please review again and approve.

skarekrow reviewed (dismissed):
Still LGTM, @sonicaj?

nathanjrobertson committed:
More travis fussiness

nathanjrobertson commented:
@skarekrow Updated to pass travis tests (whitespace warnings only).

nathanjrobertson commented:
@sonicaj Could you please take a look at this and consider merging it? I realize releases of iocage are infrequent, and that merging won't make it into FreeBSD's ports collection for a fair while, but every time there's a package update it reminds me that having to patch manually all the time is getting a bit old. And I really don't want this to miss the next release, because I've got production servers which rely on this patch. Thanks.

sonicaj commented on iocage_lib/ioc_json.py:
```diff
@@ -868,6 +868,13 @@ def check_config(self, conf, default=False):
+        for x in range(0, 4):
+            if not conf.get(f"vnet{x}_mtu"):
+                conf[f"vnet{x}_mtu"] = 'auto'
+        if not conf.get(f"vnet_default_mtu"):
```
This does not have to be a `f-string`

nathanjrobertson replied:
Yep, fair point. Got a bit keen on the copy/paste. I've pushed a patch for this one to only use f-strings where there is variable substitution.

sonicaj commented on iocage_lib/ioc_start.py:
```diff
@@ -1500,7 +1502,9 @@ def find_bridge_mtu(self, bridge):
         if not memberif:
-            return '1500'
+            if self.unit_test:
+                return "1500"
```
Can you please elaborate why we have this ? Wouldn't the default be `1500` and if no, don't you think this should be handled in the unit test rather then here ?

nathanjrobertson replied:
I'll take a look at this further. The patch changes the default to be set by a variable, but that variable doesn't get set in unit_test mode. There were other places in the code where this stubbing out for test mode happened, so it seemed to be the way it was done. It didn't seem right to me, but seemed to be a convention. Anyway, leave it with me and I'll come up with something more elegant, because I didn't like that when I did it, so I'm glad there is a better way of handling unit_test mode.

In the mean time, it seems that Cirrus CI is failing on installing packages (ie. before our tests start).

nathanjrobertson replied:
@sonicaj Ok, I've reproduced the error when I remove that "if self.unit_test: return "1500". It needs that because otherwise we get

> AttributeError: 'IOCStart' object has no attribute 'get'

in when running test mode (tests/unit_tests/1000_lib_start_test.py:55). The patch I'm looing to have merged makes the default MTU configurable, and that gets done with an IOCStart.get() call. So, I can see three options:

1. Leave it as is, with a hardcoded default that only services the unit test. This is done 30 lines above in the same function (dealing with DHCP being unavailable in unit test mode).
2. Remove the test. The test is hardcoded to test for "1500", and hardcoding the return value in #1 has very limited value in testing.
3. Remove the feature of having a default MTU other than 1500. The value of the default used to be hardcoded to 1500 (in fact, the whole thing was hardcoded to 1500 without this patch), so having a hardcoded default as long as you can configure per-interface is still useful.

I'd personally recommend solution #2. Remove what is a test which has limited value, particularly given the default can be changed. Are you ok with that being the resolution?

nathanjrobertson replied:
Actually, please don't do #3. Having a default MTU setting is actually more useful than having per-interface. Chances are, in a given environment, it'll be the same for all interfaces, so you'll end up setting each and every interface to whatever the value you need is. Possible, but more painful.

nathanjrobertson replied:
@sonicaj I've just committed an update which I'd like reviewed. I've provided a dummy IOCStart.get() for that test. So, the test remains, and the test does something useful. Could you please review again?

I'm just waiting for Cirrus-CI to finish executing. I needed to update it to an 11.3-STABLE snapshot in order to get it to execute. The version of the packages in the FreeBSD package repo are for 11.3, and don't install on 11.2 any more, it seems.

nathanjrobertson replied:
Ok. Test is fixed so that it does test the functionality, and the CirrusCI tests pass. I'll resolve this thread.

sonicaj reviewed (commented):
Thank you @nathanjrobertson, a few minor nits

nathanjrobertson committed:
Merge branch 'master' of https://github.com/iocage/iocage

nathanjrobertson committed:
Update ioc_json.py

Remove a couple of f-strings that didn't have variables within them.

nathanjrobertson committed:
Remove self.unit_test hack for vnet_default_mtu

- Instead of having dummy data in the implementation, hack the test to implement the needed IOCStart.get() method fragment we need to test.
- Also upgrade the cirrus.yml to use FreeBSD 11.3 rather than 11.2 - hopefully this fixes the failing unit test

nathanjrobertson committed:
Update cirrus-ci to FreeBSD 11.3-STABLE

nathanjrobertson committed:
Revert "Update cirrus-ci to FreeBSD 11.3-STABLE"

This reverts commit 3473e6b33fa4a5530330665e5950543687caa8b0.

nathanjrobertson committed:
Lets see if the latest 11.3 snapshot works with CirrusCI

According to https://cirrus-ci.org/guide/FreeBSD/ the 11.3-RELEASE doesn't work with Cirrus-CI / gCloud. But apparently the stable snapshots do. Lets see.

nathanjrobertson committed:
11.2 is EOL. Use 11.3 in jail.

nathanjrobertson commented:
@sonicaj I've made the changes requested, and have the CirrusCI tests running as well (they were broken across the project - the breakage wasn't related specifically to the changes I made). In fixing that, I had to update us to 11.3-RELEASE in the jail and a snapshot of 11.3-STABLE to get the tests running again (apparently CirrusCI uses gCloud, which doesn't currently work with 11.3-RELEASE, but does with -STABLE snapshots). The PR should be ready to consider for merging now.

nathanjrobertson commented:
@skarekrow @sonicaj Could somebody please take another look at this and reconsider for merging?

nathanjrobertson committed:
Use the latest 11-STABLE family image, not a specific dated snapshot (thanks to asomers for the tip)

sonicaj reviewed (approved):
Thank you @nathanjrobertson this looks good to me.
